### DISCUSSION TITLE
[Bug]: reset ignore the validation rules from register on v7

### DISCUSSION THREAD

**Discussion by yifei-zhan** [react-hook-form/react-hook-form#6156] [Q&A]
### Version Number

7

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-v7-reset-bugs-rjvdb

### Steps to reproduce

1. Go to codeSandbox link
2. Remove default inputs and leave the input field empty
3. The error message `Error` is shown(since we register with validation-required)
4. Click the "click to reset form" button(it will call `reset` api to reset the form)
5. Remove default inputs and leave the input field empty
6. The error message is not shown

### Expected behaviour

The error message should still be shown even I reset the form.
I guess the reason is the during the `reset`, the form element is registered again without validation.

**This behaviour differs from version6 (please just change the `react-hook-form` version to 6 in the sandbox and then follow the "steps to reproduce", the behaviour is as expected).**

We had a lot of dynamic forms, custom controlled input components in our project and there it is handy to use `register` to do the form logic with the `react-hook-form`.
We just started migration from v6-v7 and found out this behaviour deviation.

- [x] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
```
useEffect(() => {
    register("firstName", {
      required: "Please don't forget me"
    });
}, [register]);
```

this is not get invoked again after `reset`, reset will clear field reference as well.

**Reply by yifei-zhan** [reply]
Yes, the `useEffect` is not invoked.
Did I use the `register` for the custom-controlled-input in a correct way?
How would you suggest that I can solve this issue?
This behaviour deviation(between v6 and v7) is at least for me kind of unexpected.

**Reply by bluebill1049** [reply]
it's safe to remove `useEffect` in your usecase. just let `register` run.

**Reply by yifei-zhan** [reply]
Thanks for the reply.
May I ask you one more question?

Is it safe to say, we can **always** `register` directly in component-function body, outside of effect?
So each render will call `register` again to ensure the validation rules will not be overwritten, because sometimes the component do not know when the form is reseted.

For me, this point is a little bit anti-intuitive, since from the naming-`register`, I would expect it should be invoked only one time during the initialization, not in every rendering).

----------------------
In our code base, we used `register` & `reset` a lot for dynamic forms.
Sometimes the `reset` happens at several component levels above the `register` component.
I'm not sure whether it is safe to move all our `register` out from `useEffect` and just put it directly in the function body.

**Reply by bluebill1049** [reply]
it's fine to run it. it's very cheap and fast to run `reigster`, however you can do with register at useEffect with your example as well, but then you need a state at useEffect to trigger re-render, eg isSubmitted.

**Reply by bluebill1049** [reply]
> I'm not sure whether it is safe to move all our register out from useEffect and just put it directly in the function body.

I think it's safe to move it out. it's just a function call, no magic in it.

**Reply by yifei-zhan** [reply]
Thanks for the reply, we will try it.

**Reply by yifei-zhan** [reply]
> however you can do with register at useEffect with your example as well, but then you need a state at useEffect to trigger re-render, eg isSubmitted

In the sandbox example, the form is not even submitted. I think it is a common use case that `reset` & `submit` does not necessarily happen together.

Would you consider some patches for this issue on your side in the future? (Since in the v6, it is fine to put `register` in `useEffect` and trigger `register` only when necessary, the validation rules will not be ignored in the new render).
